### PR TITLE
Expanded on various node roles

### LINKED ISSUE #14429: Update node.(other) docs
[Here](https://www.elastic.co/guide/en/elasticsearch/reference/2.0/modules-node.html) we only list `node.data` when there is `.master` as well.

We should reflect this in the docs as a I've seen a few people ask about this just recently.

### PR DESCRIPTION
**Pull request by markwalkom**
Fixes #14429

### CODE REVIEW

File: docs/reference/modules/node.asciidoc [hunk 1]
```diff
@@ -1,32 +1,48 @@
 [[modules-node]]
 == Node
 
-*elasticsearch* allows to configure a node to either be allowed to store
-data locally or not. Storing data locally basically means that shards of
-different indices are allowed to be allocated on that node. By default,
-each node is considered to be a data node, and it can be turned off by
-setting `node.data` to `false`.
+*elasticsearch* allows the configuration of a node to play multiple
```
Comment (nik9000) [inline docs/reference/modules/node.asciidoc]: *elasticsearch* nodes have three roles and any elasticsearch node can play one to three of those roles. By default every node performs all three roles.

File: docs/reference/modules/node.asciidoc [hunk 2]
```diff
@@ -1,32 +1,48 @@
...
+`node.data: false` in the configuration file.
 
-We can start a whole cluster of data nodes which do not even start an
-HTTP transport by setting `http.enabled` to `false`. Such nodes will
-communicate with one another using the
-<<modules-transport,transport>> module. In front
-of the cluster we can start one or more "non data" nodes which will
-start with HTTP enabled. All HTTP communication will be performed
-through these "non data" nodes.
+*master nodes* control cluster state, including changes such as node
```
Comment (nik9000) [inline docs/reference/modules/node.asciidoc]: *master nodes* participate in master elections and, when elected as the _active master node_, control cluster state. If you choose to dedicate a node to just play the *master node* role then never send requests to it. This prevents the node from getting overloaded and lets it spend 100% of its time keeping cluster state up to date. The default...

File: docs/reference/modules/node.asciidoc [hunk 3]
```diff
@@ -1,32 +1,48 @@
...
+can also be set to `node.master: false` in the configuration file.
 
-The benefit of using that is first the ability to create smart load
-balancers. These "non data" nodes are still part of the cluster, and
-they redirect operations exactly to the node that holds the relevant
-data. The other benefit is the fact that for scatter / gather based
-operations (such as search), these nodes will take part of the
-processing since they will start the scatter process, and perform the
-actual gather processing.
+*client nodes* create smart load balancers that take part in some of
```
Comment (nik9000) [inline docs/reference/modules/node.asciidoc]: *client nodes* are smart load balancers that keep a copy of the cluster state and use it to route request exactly where they need to go. They also act the *coordinating node* for requests sent to them, forwarding the requests to the data nodes and then merging the results when they come back. Remember that by default all nodes act as client nodes so all nodes can be the coordinating node for queries.

I don't understand the bit you have about `node.client:true` implying `node.master: false` and `node.data: false`. Especially in light of every node being a client node by default. Is this just some funky quirk of that setting?

File: docs/reference/modules/node.asciidoc [hunk 4]
```diff
@@ -1,32 +1,48 @@
...
+We have a cluster with 15 nodes. We decide to have 3 dedicated
+master nodes to control and maintain the cluster, 10 data nodes
+to store the indexed data and 2 client nodes for users to interact
+with the cluster.
+
+The configuration file settings for the 3 masters nodes would be
+`node.data: false` and `node.master: true`, the settings for the
+10 data nodes would be `node.data: true` and `node.master: false`,
+and for the client nodes it would be `node.data: false` and
+`node.master: false`.
```
Comment (nik9000) [inline docs/reference/modules/node.asciidoc]: Do we even need `node.client` then?

File: docs/reference/modules/node.asciidoc [hunk 5]
```diff
@@ -1,32 +1,48 @@
...
+default is `node.client: true` and can also be set to `node.master: false`
+in the configuration file. A node with `node.client: true` also implies
+that `node.master: false` and `node.data: false` is set, even if not
+specified in the configuration file.
 
-This relieves the data nodes to do the heavy duty of indexing and
-searching, without needing to process HTTP requests (parsing), overload
-the network, or perform the gather processing.
+Lets take an example:
+We have a cluster with 15 nodes. We decide to have 3 dedicated
```
Comment (nik9000) [inline docs/reference/modules/node.asciidoc]: I wonder about this example. 3 dedicated master node is very typical. 2 dedicated clients isn't. Maybe its worth mentioning that?

File: docs/reference/modules/node.asciidoc (hunk 3 above)
Comment (markwalkom) [inline docs/reference/modules/node.asciidoc]: Will remove the `node.client:true` to keep it simple

File: docs/reference/modules/node.asciidoc (hunk 5 above)
Comment (markwalkom) [inline docs/reference/modules/node.asciidoc]: I think this is kinda irrelevant, this just an example to show how you can build a cluster of a given set of nodes with specific roles.

**Comment by clintongormley** [comment]
@bleskes you mentioned that you wanted to remove documentation of `node.client` in favour of just `node.data: false`, `node.master: false`.

**Comment by clintongormley** [comment]
@markwalkom just on formatting, I'd use a definition list, eg:

    `node.master`::
    +
    --
    Master nodes ... blah bah 

    Another para....
    --

    `node.data`::
    +
    --
    Data nodes ... blah bah 

    Another para....
    --

**Comment by bleskes** [comment]
>  @bleskes you mentioned that you wanted to remove documentation of node.client in favour of just node.data: false, node.master: false.

@clintongormley yeah, I think we should. I'm good with removing it here, or at least mark them as deprecate and remove the "smart balancers" wording. Note though that it takes more work. We obviously need to change code and also re-write this page (and some more where NodeBuilder is mentioned) https://github.com/elastic/elasticsearch/blob/master/docs/java-api/client.asciidoc

**Comment by markwalkom** [comment]
@clintongormley updated with the list :)

File: docs/reference/modules/node.asciidoc [hunk 6]
```diff
@@ -1,32 +1,63 @@
...
-processing since they will start the scatter process, and perform the
-actual gather processing.
-
-This relieves the data nodes to do the heavy duty of indexing and
-searching, without needing to process HTTP requests (parsing), overload
-the network, or perform the gather processing.
+*elasticsearch* nodes have three roles and any node can play
+one to three of those roles. By default every node performs all
+three roles; _data node_, _master node_ and _client node_.
+
```
Comment (clintongormley) [inline docs/reference/modules/node.asciidoc]: I'd add a line saying that the default is master & data true

File: docs/reference/modules/node.asciidoc [hunk 7]
```diff
@@ -1,32 +1,63 @@
...
+`node.master`::
++
+--
+*master nodes* participate in master elections and, when elected as
+the _active master node_, control cluster state, including changes
+such as node addition or removal, index creation and deletion and
+mapping updates.
+
+If you have dedicated master nodes then you should
```
Comment (clintongormley) [inline docs/reference/modules/node.asciidoc]: ..dedicated master nodes (ie node.master:true and node.data:false)

File: docs/reference/modules/node.asciidoc [hunk 8]
```diff
@@ -1,32 +1,63 @@
...
+not send index or query requests to them, this ensures they can
+maintain the cluster state with little to no interruption, ensuring
+the cluster remains as stable as possible.
+
+The default setting is `node.master: true` and can also be
+set to `node.master: false` via the configuration file.
```
Comment (clintongormley) [inline docs/reference/modules/node.asciidoc]: ... or by passing `--node.master false` on the command line.

File: docs/reference/modules/node.asciidoc [hunk 9]
```diff
@@ -1,32 +1,63 @@
...
+
+`node.data`::
++
+--
+*data nodes* store indexed data locally and make it available for
+further manipulation. Storing data locally basically means that
+shards of different indices are allowed to be allocated on that node.
+
+The default setting is `node.data: true`, but can also be
+set to `node.data: false` via the configuration file.
```
Comment (clintongormley) [inline docs/reference/modules/node.asciidoc]: ... or by passing `--node.data false` on the command line.

**Comment by clintongormley** [comment]
I think we should remove mention of `--node.client` and "smart router" as it sounds too attractive.  This isn't something we recommend these days.

**Comment by markwalkom** [comment]
@clintongormley do you specifically mean `node.client` or the concept of a client node? I just want to make sure we're on the same page.

**Comment by clintongormley** [comment]
Specifically `node.client` (just set master and data to false) but also we don't usually recommend setting up client-only nodes any more.  We prefer users to use the transport client instead.

**Comment by clintongormley** [comment]
Closing in favour of c0a7f8889781f342c6c79696de904bc4ef47314b

**Pull request closed without merging**